### PR TITLE
Search bar

### LINKED ISSUE #1123: Search function for site
I usually click around a fair amount to find what I'm looking for. Often it is not clear to me what is a "style" or a "component." For example I would think of a "logo" or "breadcrumbs" as components that make up the structure of the site, but they are considered styles. I'm know the Cashmere team surely has some clear delineations defined, but a search may be more helpful than expecting every user to learn and understand them. I do understand that search is possible in the GitHub interface. I am not a UI/UX or graphic designer, and I do understand that people in those roles are who the site is primarily for, so most users in those areas of expertise may inherently understand the delineations.  I use the site a fair amount to make sure things I work on align with Cashmere standards or to pull resources off. Thanks for considering.

### LINKED ISSUE #1351: Deep linking support for all of the Cashmere documentation
In order to implement #1123 , we first need a way to navigate to any piece of content on the site.  We can accomplish this by adding `id` attributes to header tags, for example:

```
<h2 id="some-header">Some header</h2>
```
This will allow navigating to the header using the hash `#some-header` at the end of the URL.
You can play around with this idea here: <https://stackblitz.com/edit/angular-heading-hash-links?file=src/app/app.component.html>

Where headers are unavailable, we can use anchors (`<a name="some-section"></a>`) to do the same thing.

Tasks for completing this:

- [x] Create a deep link to every example on each component page (see `example-viewer.component.html`)
- [x] Create a deep link to every tab in the example (TS, HTML, CSS, etc)
        _This could be a stinker_
- [x] Create a deep link for every header rendered from Markdown
        Currently we are using `markdown-it` for rendering Markdown to HTML, and it does not support rendering ids or anchors in headers.  Alternatively, the `marked` package supports GFM and rendering ids in headers, so we should explore replacing the `markdown-it` library with `marked` (https://github.com/markedjs/marked) - Marked GitHub stars
- [x] Create a deep link for headers within the generated API documentation - already done :)
        ~We use dgeni (<https://github.com/angular/dgeni>) for generating our API documentation from source.  It may be tricky, but we need to figure out how to get the proper headings and/or anchors inserted into the generated docs.  If this isn't possible, we may need to consider an alternative code documentation generator (like TypeDoc or something).~
        Luckily, dgeni already does this for us!
- [x] Fix the CSS for the Cashmere site to scroll under the navbar, instead of to the right of the navbar.  Otherwise the header gets cut off by the navbar when you go to the hash link.

### LINKED ISSUE #1363: Component and Guides routing bug
When on the Components or Guides page, clicking on the link in the navbar clears the page instead of taking the user to the first topic on the page. For example, normally clicking on components would result on the user being routed to https://cashmere.healthcatalyst.net/components/checkbox/examples but once the user clicks on the link again, they are routed to https://cashmere.healthcatalyst.net/components

Steps to reproduce:

1. Click on the Components or Guides tab in the navbar
2. Click on the same tab again and the user should be routed to the wrong URL.

Image of blank page -
image

Image of the URL - 
image

### PR DESCRIPTION
**Pull request by andrew-frueh**
This one is about ready for prime-time!  Thanks to @CameronAFaust and @CristianIxcamey for all their work getting the ball rolling on this one.  I think I've got the search indexing, deep-linking, etc all finalized.  @joeskeen and @corykon , I'll demo the changes for you on our call tomorrow.  

Lots of files changed here because there were some updates made to markdown or style guides to ensure everyone conformed to the same standards.

closes #1123 
closes #1351
closes #1363

### CODE REVIEW

**Review by corykon — APPROVED** [review]
Looks great, stoked for this. I didn't dig deeply into the code itself, just played with the search UI.
Let's get it out there and get some feedback.

File: src/app/styles/logo/logo-demo.component.ts [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
     styleUrls: ['./logo-demo.component.scss']
 })
-export class LogoDemoComponent {}
+export class LogoDemoComponent implements AfterViewInit {
+    private section: string | null;
+
+    constructor(private router: Router) {}
+
+    ngAfterViewInit() {
+        this.section = this.extractUrlValue( 'section', this.router.url );
```
Comment (joeskeen) [inline src/app/styles/logo/logo-demo.component.ts:15]: This function exists everywhere... couldn't we DRY it up by extracting it out?

File: .vscode/settings.json [hunk 2]
```diff
@@ -4,6 +4,7 @@
     "editor.tabSize": 4,
     "typescript.tsdk": "node_modules\\typescript\\lib",
     "cSpell.words": [
-        "Strat"
+        "Strat",
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply .vscode/settings.json:7]: Beats me where this came from - removing.

File: scripts/generate-usage-docs.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-const md = require('markdown-it')({html: true});
+const mark = require('marked');
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply scripts/generate-usage-docs.js:1]: Pretty sure this was an abandoned attempt by the team to get ids added to the usage docs.  I wound up doing it another way, so I'm going to roll back the changes to this file.  Great catch.

File: src/app/styles/logo/logo-demo.component.ts (hunk 1 above)
Comment (joeskeen) [reply src/app/styles/logo/logo-demo.component.ts:15]: Perhaps a small SectionScrollService?  The service could listen to the Router's URL change to store the current `section`, then the component calls `scrollToSection()` in `ngAfterViewInit`?

File: src/app/app.component.ts [hunk 4]
```diff
@@ -1,8 +1,59 @@
...
+        'bits': { icon: 'fa-puzzle-piece' }
+    };
+
+    constructor( private searchService: SearchService ) { }
+
+    ngAfterViewInit() {
+        this.navSearchBar.valueChanges.subscribe((val) => {
+            if (val !== '') {
+                let tempResults = this.getItems(val);
+                this.searchResults = tempResults.slice(0, 4);
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply src/app/app.component.ts:34]: I have a feeling that's what I had in my prototype, so the team took it literally :-)  I'm with you, I'll switch it to 5.

File: src/app/components/component-viewer/component-viewer.component.ts [hunk 5]
```diff
@@ -18,20 +21,31 @@ export class ComponentViewerComponent {
...
 
         if (this.docItem) {
             const examples = this.docItem.examples;
+            const urlArray = this.router.url.split('/');
             if (examples && examples.length > 0) {
                 availableSections = ['Examples'];
             }
             if (!this.docItem.hideApi) {
                 availableSections.push('API');
+                if ( urlArray[ urlArray.length - 1 ].substr(0, 3) === 'api' ) {
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply src/app/components/component-viewer/component-viewer.component.ts:38]: Agreed, good call.  Will change.

File: src/app/components/components.component.ts [hunk 6]
```diff
@@ -62,22 +61,27 @@ export class ComponentsComponent implements OnInit, OnDestroy {
...
 
         this.allDocItems = this.docItemService.getDocItems(this.docType);
-        if (!this.id && this.categorizedDocItems[0] && this.categorizedDocItems[0].items && this.categorizedDocItems[0].items.length) {
-            this.navUpdate(this.categorizedDocItems[0].items[0].id);
-            return;
+        if (!this.id) {
+            for ( let i = 0; i < this.categorizedDocItems.length; i++ ) {
+                const tempVal = this.categorizedDocItems[i].items;
+                if ( tempVal && tempVal.length ) {
+                    this.navUpdate(tempVal[0].id);
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply src/app/components/components.component.ts:77]: Let's get your Bit PR merged first, and I'll resolve the conflict to keep your change.  Looks like we were both trying to fix a few of the same bugs.

File: src/app/core/cashmere-components-document-items.json [hunk 7]
```diff
@@ -19,7 +19,7 @@
     "button": {
         "name": "Button",
         "category": "buttons",
-        "examples": ["button-anchor", "button-colors", "button-icon", "button-link", "button-size", "button-split", "button-type"],
+        "examples": ["button-type", "button-anchor", "button-colors", "button-icon", "button-link", "button-size", "button-split"],
```
[the remark this answers is not in the document]
Comment (andrew-frueh) [reply src/app/core/cashmere-components-document-items.json:22]: Yes, that was my change - types was buried at the bottom, but it's the most basic example, so I wanted to bump it up to the top.

**Review by joeskeen — APPROVED** [review]

**Pull request merged**